src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs

```diff
@@ -223,88 +221,56 @@ internal void ForceCancelRequest()
 
     /// <summary>
     /// Attempts to get the client hello message bytes from the http.sys.
-    /// If not successful, will return false.
+    /// If successful writes the bytes into <paramref name="destination"/>, and shows how many bytes were written in <paramref name="bytesReturned"/>.
+    /// If not successful because <paramref name="destination"/> is not large enough, returns false and shows a size of <paramref name="destination"/> required in <paramref name="bytesReturned"/>.
+    /// If not successful for other reason - throws exception with message/errorCode.
     /// </summary>
-    internal unsafe bool GetAndInvokeTlsClientHelloMessageBytesCallback(IFeatureCollection features, Action<IFeatureCollection, ReadOnlySpan<byte>> tlsClientHelloBytesCallback)
+    internal unsafe bool TryGetTlsClientHelloMessageBytes(
+        Span<byte> destination,
+        out int bytesReturned)
     {
+        bytesReturned = default;
         if (!HttpApi.SupportsClientHello)
         {
             // not supported, so we just return and don't invoke the callback
-            return false;
+            throw new InvalidOperationException("HttpApi does not support client hello. See HTTP_FEATURE_ID.HttpFeatureCacheTlsClientHello");
```

Comment:
Should this be a `PlatformNotSupportedException`? Also, it might help clarify that this isn't a limitation of HTTP.Sys in general, but this specific version of the API.

```suggestion
            throw new PlatformNotSupportedException("This version of the Windows HTTP Server API does not support client hello. See HTTP_FEATURE_ID.HttpFeatureCacheTlsClientHello");
```

Also, do we expect people using this API are going to have to query support using the HTTP_FEATURE_ID? I think it's fine information to include in the exception message, but if the consumers of the API actually have to check this, we might need to expose a capability API in .NET so people don't have to p/invoke. If the expectation is that this won't be called from other library code, and that anyone calling this knows they'll be deploying the code only to servers that support this API, then I think we're fine without the capability API.

Reply:
> ```
>     /// Works only if <c>HTTP_SERVICE_CONFIG_SSL_FLAG_ENABLE_CACHE_CLIENT_HELLO</c> flag is set on http.sys service configuration.
>     /// See <see href="https://learn.microsoft.com/windows/win32/api/http/nf-http-httpsetserviceconfiguration"/>
>     /// and <see href="https://learn.microsoft.com/windows/win32/api/http/ne-http-http_service_config_id"/>
> ```

I just read this is the doc comment. Does that mean that this will throw by default if you haven't opted into the feature via configuration? In that case, I think the InvalidOperationException is correct, but I'd include this bit from the doc comment in the exception message as well.

Reply:
> Does that mean that this will throw by default if you haven't opted into the feature via configuration?

`HttpApi.HttpGetRequestProperty` will be throwing if flag is not set on http.sys level.

However, we can not check if `HTTP_SERVICE_CONFIG_SSL_FLAG_ENABLE_CACHE_CLIENT_HELLO` was set (requires more infra setup and discovery), so it is out of aspnetcore's control to check it. `HttpApi.SupportsClientHello` only validates the support of `HTTP_FEATURE_ID.HttpFeatureCacheTlsClientHello`, therefore including that in the exception message: 
> Windows HTTP Server API does not support HTTP_FEATURE_ID.HttpFeatureCacheTlsClientHello or HttpQueryRequestProperty. See HTTP_FEATURE_ID for details.